aqueduct/worker.py

```diff
@@ -75,57 +41,106 @@ def _start(self):
         """Runs something huge (e.g. model) in child process."""
         self.task_handler.on_start()
 
-    def _tasks(self) -> Iterator[Optional[BaseTask]]:
-        """Provides suitable for processing tasks."""
-        while True:
-            try:
+    def _get_task(self, timeout: float):
+        try:
+            if timeout == 0:
                 task: BaseTask = self.queue_in.get(block=False)
-            except queue.Empty:
-                # returns control
-                yield
-                time.sleep(0.001)
-                continue
-            if isinstance(task, StopTask):
-                self._stop_task = task
+            else:
+                task: BaseTask = self.queue_in.get(timeout=timeout)
+        except queue.Empty:
+            return
+
+        if isinstance(task, StopTask):
+            self._stop_task = task
+            return
+
+        log.debug(f'[{self.name}] Have message')
+        task.metrics.stop_transfer_timer(self.step_name)
+        # dont't pass an expired task to the next steps
+        if task.is_expired():
+            log.debug(f'[{self.name}] Task expired. Skip: {task}')
+            return
+
+        # don't process unsuitable tasks
+        if not self.handle_condition(task):
+            self._post_handle(task)
+            return
+
+        return task
+
+    def _get_batch_with_timeout(self, batch):
+        timeout = self._batch_timeout
+
+        while True:
+            with timeit() as passed_time:
+                task = self._get_task(timeout)
+
+            if task:
+                batch.append(task)
+
+                if len(batch) == self._batch_size:
+                    return batch
+
+            timeout -= passed_time.seconds
+
+            if self._stop_task or timeout <= 0:
+                return batch
+
+            timeout = max(timeout, 0.1)
+
+    def _get_batch_dynamic(self, batch):
+        while True:
+            task = self._get_task(timeout=0)
+            if task:
+                batch.append(task)
+                if len(batch) == self._batch_size:
+                    break
+            else:
                 break
-            log.debug(f'[{self.name}] Have message')
-            task.metrics.stop_transfer_timer(self.step_name)
-            # dont't pass an expired task to the next steps
-            if task.is_expired():
-                log.debug(f'[{self.name}] Task expired. Skip: {task}')
-                continue
-            # don't process unsuitable tasks
-            if not self.handle_condition(task):
-                self._post_handle(task)
-                continue
-            yield task
 
-    def _tasks_batches(self) -> Iterator[Optional[List[BaseTask]]]:
+        return batch
+
+    def _get_batch(self):
+        batch = []
+
+        # wait first task
+        while True:
+            task = self._get_task(10.)
```

Comment:
Please add a comment why do  we use exactly 10 seconds

Reply:
added comment